crates/cargo-gpu/src/main.rs

```diff
@@ -63,7 +65,37 @@ mod spirv_cli;
 mod spirv_source;
 mod toml;
 
-fn main() {
+/// Central function to write to the user.
+#[macro_export]
+macro_rules! user_output {
+    ($($args: tt)*) => {
+        #[allow(
+            clippy::allow_attributes,
+            reason = "`std::io::Write` is only sometimes called??"
+        )]
+        #[allow(
+            clippy::useless_attribute,
+            reason = "`std::io::Write` is only sometimes called??"
+        )]
+        #[allow(
+            unused_imports,
+            reason = "`std::io::Write` is only sometimes called??"
+        )]
+        use std::io::Write as _;
+
+        #[expect(
+            clippy::non_ascii_literal,
+            reason = "CRAB GOOD. CRAB IMPORTANT."
+        )]
+        {
+            print!("🦀 ");
+        }
+        print!($($args)*);
+        std::io::stdout().flush().unwrap();
+   }
+}
+
+fn main() -> anyhow::Result<()> {
```

Comment:
In general I think it's fine for a cli program to `panic` when it hits an unrecoverable error. There are benefits to this behavior like knowing exactly where in the program the failure occurred since the line number is included in the print statement and a backtrace is produced. 
But wrapping all possible errors in `Result` can be nice for testing and is a generally good practice. 
So - my only concern here is that if `main` returns the `Result`, we have no way of logging the error (and the errors further up the chain of errors) or the backtrace and instead we rely on whatever Rust's default behavior is - which I think is likely just panicking with the text of the top-level error, erasing the backtrace. 

So I think here we should probably not return `Result` and instead `if let Err(msg) = result { ... }` making sure to use all of `anyhow`'s machinery to print the relevant information.

Reply:
Good points, I agree. I've refactored so that `main()` wraps and captures the error from a `run()` function, then writes the error message to `stderr` and logs the backtrace when not a release build.